Advanced Routing (#1344)
withastro/roadmap: proposals/0056-advanced-routing.md

- Start Date: 2026-04-21
- Reference Issues: https://github.com/withastro/roadmap/discussions/1320
- Implementation PR: https://github.com/withastro/astro/pull/16366
- Stage 2 Issue: https://github.com/withastro/roadmap/issues/1342
- Stage 3 PR: https://github.com/withastro/roadmap/pull/1344

# Summary

Provides greater control over the request lifecycle in Astro as well as the ability to use server frameworks that use the Request/Response APIs such as [Hono](https://hono.dev/), allowing the user to inject logic between Astro features such as i18n, redirects, rewrites, sessions, and even the pages directory.

# Example

> The minimal API for this proposal is a `src/app.ts` file with the following shape:

Reviewer: At this point I would call it `src/fetch.ts`. Everything orbits around a public-facing fetch APIs and concepts. The only time we mention "app", is when we mention the `App` class, which is and stays an internal implementation. 

Also, highlight the fact that the file can be any supported extension `.js`, `.mjs`, etc.

```ts
export default {
  fetch(request: Request) {
    return new Response('ok', {
      status: 200
    });
  }

> }

Reviewer: I think it's fine to keep the example like this, however I think we should expose a `defineFetch` (or similar) utility like we do for actions (`defineAction`), middleware (`defineMiddleware`) and so on. This will improve type-safety. For examples, things might not be clear:
- `fetch` is `async` by design
- `fetch` must return a `Response`

Author: I added an explanation of the shape and why it's that way here: [239213b](https://github.com/withastro/roadmap/pull/1344/commits/239213b02740d28f9edcb0e395d5b7f28918cdc1)

Added a use of `satisfies Fetchable` as that aligns with how other platforms do type safety.

- **Features are tightly coupled to Astro internals.** Middleware, Actions, rewrites, and i18n are all baked into Astro's request handling. They can't be used independently, reordered, or replaced -- they run when and how Astro decides.
- **Hard to integrate non-Astro request logic.** If you want to mount an API from another library, add rate limiting, or run platform-specific logic before Astro sees the request, you're working against the framework rather than with it.
- **Not all requests reach user code.** Requests that do not match a RouteData object never get passed through Astro's pipeline or reach user middleware, leading users to add logic in front of Astro through their platform's own APIs.

This proposal aims to consolidate a single pipeline for Astro request handling where all requests go into, and the user has complete control.

# Goals

- Allow developers to have complete control over requests coming into their Astro application.
- Provide an API that is completely compatible with Fetch handlers, such as Hono, so users can gain the benefit of those ecosystems.

> - Break individual Astro features into small APIs that can be composed however the user wishes.

Reviewer: I think it's important to state if the goal is to have this feature working for SSG, SSR, both, or some of them.

Author: Added a new goal here to clarify it works the same in SSG and SSR.

# Non-Goals

- Creating a single "entry point" that covers all platform-specific APIs. This API still runs after adapters.
- Removing any features; this is just a different way to control what order features run. Features such as injectRoute will still exist, even if it would be possible to implement similar functionality through Hono middleware.
- Provide ways for integrations to inject code into the pipeline. The user has complete control over the request pipeline in this proposal; in fact they have to include the right middleware to even render pages.

# Detailed Design

Currently most of the logic to resolve what gets called during a request is contained within the App class. This class doubles as the external API used by adapters to render pages.

The bulk of the changes for this proposal will be to extract the logic out of the App class and into feature-specific handler classes. Each feature is organized as a class, but the API of each handler varies depending on what it does. There are a few common patterns:

> - **Request interceptors** check the request and may return a response early (e.g. redirects). These have a `handle(state: FetchState)` method that returns a `Response` or `undefined`:

Reviewer: Two things need to be addressed imho:
- Calling it _request_ interceptor might be misleading because `handle` accepts a `FetchState`, not a `Request`. 
- It's important to highlight if `state` is a copy or not, and if `state` is meant to be mutable in an imperative way

Author: **Request interceptors** seems approach here because it's meant to describe the pattern and what these handlers do is intercept the request and return responses if the request matches what it is that they handle.

Remember that this is an explanation of the internal code, not public documentation. 

Also, I updated the text to explain that FetchState is mutable here: 6346abb

```ts
class RedirectsHandler {
  handle(state: FetchState): Response | undefined {
    // ...
  }
}
```

> - **Post-processors** run after the main response has been produced and may mutate it (e.g. i18n rerouting unhandled requests). These expose a method like `finalize()`.

Reviewer: nit: Add code snippet for the finalize function, like you did for `handle`. Also, explain what "main response" means from an architectural point of view. Where is it coming from? What creates it? etc.

Author: Added this here: b2d9e9e

- **Helpers** provide checks or utilities that other handlers use, without directly producing or modifying responses.

This proposal doesn't aim to enforce a single interface across all handlers, as the methods vary depending on what each feature needs to do.

The user-facing API in `astro/fetch` wraps these classes as plain functions (e.g. `redirects(state)`, `pages(state)`). The `astro/hono` API wraps them further as Hono middleware. See the [Feature Handlers](#feature-handlers) and [Hono API](#hono-api) sections below for details.

> Internally, handler classes require configuration from Astro's SSR manifest (route table, i18n settings, etc.). To keep the user-facing API simple, the `astro/fetch` module imports the manifest via a Vite virtual module (`virtual:astro:manifest`) at build time and passes it into the handler constructors. This means users just call `redirects(state)` rather than needing to import and wire up the manifest themselves. The handler classes themselves remain pure and accept the manifest as a constructor argument, which makes them directly unit-testable without the virtual module.

Reviewer: This is a very important detail that it might shape the APIs. All APIs depend on a vite virtual module, which means that they can't be used anywhere outside of `src/app.ts`, and require vite. Given that, I do wonder if this should be a virtual module `astro:fetch` at this point.

Author: Yeah good point, I added [21edb9b](https://github.com/withastro/roadmap/pull/1344/commits/21edb9bf96931f7e66e3a08f2b9704f7149c7be4) to address this, the reason for `astro/fetch` is the idea that people will want to import the handlers to create higher-level handlers; for example Starlight might import all of them to do its own opinionated rendering logic. For that you want a real module if you're going to publish to node_modules/ while depending on this API.

Reviewer: Right! Didn't think of that. I suppose we would need to document that users need vitest to test their fetch handler

## FetchState

> Every request has additional state associated with it. Some of it is specific to the adapter such as the `clientAddress`, while other state is loaded during the request lifecycle, such as the `APIContext` object that is created to pass to user middleware.

Reviewer: ```suggestion
Every request has additional state associated with it. Some of it is specific to the adapter such as the `clientAddress`, while other state is loaded during the request lifecycle, such as the `APIContext` object that is created to pass to user middleware or endpoint.
```

> Every request has additional state associated with it. Some of it is specific to the adapter such as the `clientAddress`, while other state is loaded during the request lifecycle, such as the `APIContext` object that is created to pass to user middleware.

Reviewer: Also explain if `state` is meant to be mutable or not.

> The `FetchState` object is the representation of this state and needs to be created by the user when using the `astro/fetch` API. The expected public surface includes:

Reviewer: I think I saw `locals` too during the API bash.

Author: Added that here: be415c7

> The `FetchState` object is the representation of this state and needs to be created by the user when using the `astro/fetch` API. The expected public surface includes:

Reviewer: How do users have access to the manifest/configuration?

- `request` - The current `Request` object.

> - `response` - The `Response` produced by handlers, if any.

Reviewer: It's important to mention where a `response` is coming from.

- `routeData` - The matched route information.

> - `pathname` - The resolved pathname for the request.

Reviewer: Important to highlight that `pathname` may differ in build e.g. `.html` extension

Author: Added a note that pathname may include a `.html` extension during static builds.

- `cookies` - The `AstroCookies` instance for reading/writing cookies.

> **Note:** The exact shape of `FetchState` is still a work in progress and may change during the review and implementation process.

Usage:

```ts
import { FetchState, pages, redirects } from 'astro/fetch';

export default {
  fetch(request: Request) {
    const state = new FetchState(request);

    // Run the redirects
    const response = redirects(state);
    if(response) {
      return response;

>     }

Reviewer: So, you mentioned that `state` might have a `response`. Given that `redirects` returned a `response`, I do wonder which one should be the source of truth.

Author: I'm going to see if we really need `response` on FetchState and if not, remove it.

    // Continue...
    return pages(state);
  }
}
```

Note that using the `astro/hono` API the creation of FetchState is not necessary, since Hono has its own Context object; handlers can get/create a FetchState by inspecting HonoContext for a certain key (likely `astro.fetchState`).

### Context Providers

`FetchState` includes a **provider registry** that allows handlers to lazily contribute values to the `APIContext` and `Astro` global. This is the mechanism that powers features like sessions and cache — instead of being created eagerly during request setup, they are registered as providers and only instantiated when user code accesses them.

```ts
interface ContextProvider<T> {
  /** Factory called lazily on the first access. */
  create: () => T;
  /** Optional cleanup / persist callback. */
  finalize?: (value: T) => Promise<void> | void;
}
```
The API on `FetchState`:

- `provide<T>(key, provider)` — Registers a provider. The `create` factory is deferred until the first time accessed.
- `resolve<T>(key)` — Lazily calls `create()`, caches the result, and returns it. Returns `undefined` if no provider was registered.

> - `finalizeAll()` — Runs all registered `finalize` callbacks for providers that were actually resolved. Returns synchronously when nothing needs finalizing.

Reviewer: Please state if these functions are `async` by design or not

Author: Updated the signatures with explicit return types. `provide` and `resolve` are synchronous, `finalizeAll` returns `Promise<void> | void`.

For example, the `sessions()` handler registers a provider that lazily creates an `AstroSession` and persists it on finalize:

```ts
state.provide<AstroSession>('session', {
  create() {
    return new AstroSession({ cookies, config, ... });
  },
  finalize(session) {
    return session.persist();
  },
});
```

> From the user's perspective they only need to add the `session()` handler/middleware and then they get `ctx.session` / `Astro.session` as values. This could also be used by 3rd party integrations.

Reviewer: How do types work here? Let's suppose I create a fetch handler **in my project**, which returns `new OAuthSession()` from the `create` function, with key `oauth`, how can I make sure that `ctx.oauth` is properly typed when implementing my endpoint/middleware?  

For example, for `locals` users need to extend the types themselves. How does it work in this case?

Author: Same pattern as `locals`. Added a "Typing Context Providers" section showing how users extend `App.Providers` in `env.d.ts` to type custom keys like `ctx.oauth`. Built-in providers like `session` are pre-declared by Astro.

```ts
function actions(state: FetchState): Promise<Response | undefined> | undefined
```

### `sessions(state)`

Registers the session provider on the state. The session object is created lazily — only when user code accesses `ctx.session` or `Astro.session`. No-op if sessions are not configured in `astro.config`.

Call this early, before middleware runs. After the response is produced, call `state.finalizeAll()` in a `finally` block to persist any session mutations.

```ts
function sessions(state: FetchState): Promise<void> | void
```

Returns `undefined` when sessions are not configured.

### `middleware(state, next)`

Runs the user's `src/middleware.ts` (if present) around the `next` callback. The user's `onRequest` handler receives the usual `APIContext` and a `next` function. If no user middleware is configured, calls `next` directly.

```ts
function middleware(
  state: FetchState,

>   next: (state: FetchState) => Promise<Response>,

Reviewer: Is this one of the cases where `state.response` is populated? If that's the case, is `response` the same response returned by the `next` function?

Author: Yes, `middleware` sets `state.response` to the Response returned by `next`, and they're the same object. Added a clarification with a `response === state.response` comment in the code example.

): Promise<Response>
```

The `next` callback is where you put the inner handlers (typically `pages`):

```ts
const response = await middleware(state, (s) => pages(s));
```

### `pages(state)`

Renders the matched page or endpoint and returns the response. This is the core rendering handler — it loads the route's component module, executes it, and builds the response. Returns 404/500 error pages when appropriate.

```ts
function pages(state: FetchState): Promise<Response>
```

### `i18n(state, response)`

Post-processes a response against the app's i18n configuration. Handles locale redirects for the root URL, 404 responses for invalid locales, and fallback routing. Returns the response unmodified if i18n is not configured or the strategy is `'manual'`.

```ts
function i18n(state: FetchState, response: Response): Promise<Response>
```

> Call this after the response has been produced (typically by `middleware` + `pages`).

Reviewer: ```suggestion
This API requires a `Response`, you must call it after a fetch handler that returns a `Response` object, for example `middleware` or/and `pages`.
```

### `cache(state, next)`

Wraps a render callback with cache provider logic. Handles runtime caching (providers that implement `onRequest`), CDN-based providers (headers only), and the no-cache case transparently. Cache headers are applied and stripped internally.

```ts
function cache(
  state: FetchState,

>   next: () => Promise<Response>,

Reviewer: Why doesn't this callback accept `state`?

Author: `cache` only wraps the response, so it doesn't need to forward state into `next`. The caller already has `state` in scope and can close over it. Added a note explaining the difference from `middleware` with an example.

```ts
function astro(state: FetchState): Promise<Response>
```

Most users who don't need fine-grained control should use this.

### Composing handlers

Here's how the individual handlers compose into a full pipeline:

```ts
import {
  FetchState, trailingSlash, redirects,
  sessions, actions, middleware, pages, i18n
} from 'astro/fetch';

export default {
  async fetch(request: Request) {
    const state = new FetchState(request);

    // Early exits — return immediately if they handle the request.
    const slash = trailingSlash(state);
    if (slash) return slash;

    const redirect = redirects(state);
    if (redirect) return redirect;

    // Register session provider (lazy, no-op if unconfigured).
    sessions(state);

    const action = await actions(state);
    if (action) return action;

    try {
      // Middleware wraps page rendering; i18n post-processes.
      const response = await middleware(state, () => pages(state));
      return await i18n(state, response);
    } finally {
      // Persist sessions, etc.
      await state.finalizeAll();

>     }

Reviewer: I think it's a bit inconsistent that `middleware` and/or `i18n` requires a try/catch, while the other functions don't, but I understand we need some error handling in place. 

The problem I see is that `finalizeAll` wasn't mentioned as something that users need to call, because `provide` and `resolve` are called either by Astro, or the library authors. 

In this case `finalizeAll` is called **by the end-user**, which feels inconsistent with the design of the context providers.

I see a few options:
- error handling should be controlled by the library authors
- error handling should be controlled by Astro
- users control error handling but low-level APIs are still called by us
- another solution that is more consistent with the design

Author: The user needs to call it because they created the FetchState instance. If you use `astro()` you don't because it does that internally. But if you use the lower-level APIs then we can't call it because we wouldn't know when/where to do so, any handler can be called in any order, only the user knows when the request is finished.

If I update the RFC to explain this, does that work?

  }
};
```

> The power of this API is that you can slot your own logic anywhere. Add auth before `pages`, add logging around `middleware`, skip `i18n` entirely, or replace `pages` with your own rendering — it's all just function calls.

Reviewer: ```suggestion
The API is flexible, so it's possible to slot business logic anywhere. Add auth before `pages`, add logging around `middleware`, skip `i18n` entirely, or replace `pages` with your own rendering — it's all just function calls.
```

While I understand that this API is powerful, this comes with downsides. I suggested a more objective term, and removed the use of "you".

## Platform Entrypoints

`src/app.ts` runs within Astro's request handling, after the adapter and any platform-specific entrypoint (e.g. Cloudflare's `worker.ts`). The layering is:

1. **Adapter / platform entrypoint** (e.g. `worker.ts`) - Platform-specific logic, receives the raw platform request.
2. **`src/app.ts`** - User's fetch handler, receives a standard `Request`.

> 3. **Feature handlers** - Astro features like redirects, pages, etc.

Reviewer: We need to take edge middleware into account too.

Author: Added a note that platform edge middleware (Vercel, Netlify, etc.) runs before the adapter and is outside Astro's control. This proposal doesn't change how that layer works.

Reviewer: That's not what I meant; my question runs deeper. I'll explain. Up until now, Astro **was in charge** of the middleware, completely. When edge middleware is enabled, the entire file and business logic are placed on the edge, meaning the middleware **no longer runs within the Astro rendering pipeline**.

However, with this feature, the user is now in charge of where the middleware runs. So if the user decides to use `middleware` from `astro/fetch` or `astro/hono`, and then enables Netlify edge middleware, what happens? Does `middleware` become a no-op? Do we have two middlewares? 

It could be interesting to add a test in our adapters and see what the result is.

`src/app.ts` does not replace platform entrypoints. Users who need platform-specific APIs (e.g. Durable Objects, queues) still use their platform's entrypoint for those concerns.

## Hono API

> 

Reviewer: I think it's important that this RFC answers the following questions:
- why was `hono` chosen?
- why having a direct `astro/hono` specifier, instead of `@astrojs/hono` package? - Considering that Astro won't depend on `hono`, there's no reason for having a separate package.

Author: Added a paragraph at the top of the Hono API section explaining the choice: fetch-native, largest middleware ecosystem among fetch frameworks, and lightweight with no server runtime. The `astro/hono` vs `@astrojs/hono` question is covered in the Alternatives section.

A hono-specific API of middleware will be provided as `astro/hono`. These will be thin wrappers around the lower-level feature handlers that essentially just store FetchState on HonoContext, but otherwise just directly call into the handlers.

User-facing API will look like:

```ts
import { redirects, actions, pages } from 'astro/hono';
import { Hono } from 'hono';

const app = new Hono();
app.use(redirects());
app.use(actions());
app.use(pages());

> export default app;

Reviewer: I would like to have a more in-depth paragraph of the functions exported from `astro/hono`. From the API bash, you mentioned that these functions are small wrappers around the functions exported from `astro/fetch`. If that's still the case, a paragraph should address the following concerns/questions:
- Since `src/app.ts` runs for prerendered pages too, does that mean we spawn an Hono server while rendering them? 
- If `redirects` is a small wrapper of `redirects` coming from `astro/fetch`, does that mean that it inherits the same restrictions? E.g. vite dependent, can't be used outside Astro **and** vite.

The `src/app.ts` module will be opt-in. Without this file Astro will behave as it normally does.

Initially this feature will be gated by an `experimental.advancedRouting` flag. In the next major version of Astro there will be no flag, meaning that `src/app.ts` becomes a special-file that users can't use for other purposes.

To accommodate projects which already have a `app.ts` file, this will be configurable:

```
export default defineConfig({
  fetchFile: 'fetch.ts'
});
```

The shape will be:

```ts
type FetchFile = string | null;
```

The usage of `null` will disable the feature; this is useful if the user has their own `app.ts` file but don't want to define their own fetch handler file.

# Unresolved Questions

The exact shape of every feature handler is in-progress and not known. I think the shape of these can be discussed in the review process.

How should users call into Astro from platform-specific entrypoints? For example, a Cloudflare `worker.ts` that needs to export Durable Objects alongside Astro's handler might look something like:

```ts
import { handler } from 'astro/fetch';

export class MyDurableObject { ... }

export default {
  fetch: handler
}
```

> This proposal does not currently address this use case, but it may be something we want to support as part of or alongside this work.

Reviewer 2: I would like to bring some more light into this. Right now it looks a bit too complicated for users who use platform-specific entrypoints and advanced-routing the same time, as well as us to maintain both. If I understand everything correctly a request flow looks like this

```
-> raw request -> worker.ts -> handler(packages/integrations/cloudflare/src/utils/handler.ts) -> transformed request -> app.ts ->
```

Which would lead to the following code:

1. `src/worker.ts`

```ts
import handler from "@astrojs/cloudflare/entrypoints/server";

export default {
	// raw request from the worker
	async fetch(request, env, ctx) {
		return handler.fetch(request, env, ctx);
	},
} satisfies ExportedHandler<Env>;
```

2. `src/app.ts`

```ts
import { FetchState, astro } from 'astro/fetch';

export default {
  fetch(request: Request) {
    // transformed request which comes from the handler in @astrojs/cloudflare/entrypoints/server
    const state = new FetchState(request);
    return astro(state);
  }
}
```

Decision Matrix (x means the file is required, if both columns have ?, one **or** the other can be used.):

| use-cases | needs worker.ts | needs app.ts |
|--------------|-----------------|--------------|
| platform-specific features e.g. durable objects | x | |
| custom routing order | | x |
| platform-specific features e.g. durable objects & custom routing order | x            | x       |
| custom fetch, e.g. `/api/*`  | ? | ? |
| access bindings directly  | ? | ? |
| use hono | ? | ? |

As you can see from the above matrix, there is not really any value for users to use advanced-routing other when they need to use a custom pipeline order. If they want to add custom handling based on a url `/api/*` they can use `worker.ts`. If they want to add or use hono, they can use `worker.ts`. If they want to access bindings directly they can use `workers.ts`. For all these use-cases they can also use `app.ts`, but they should prefer `worker.ts` since it's faster and allows for future usage of durable objects, without a new file. They only reason I can think of why they should use `app.ts` is if they need to change the order of astro's pipeline.

This in my opinion has two downsides, it makes this harder to document and explain for us when to use which, and we still need to keep and maintain `packages/integrations/cloudflare/src/utils/handler.ts`. In my opinion and in an ideally perfect world, `worker.ts` get's obsolete, and `app.ts` supports platform-specific entrypoints, including raw request handling, top-level exports e.g. Durable Objects, ideally so we can remove the `packages/integrations/cloudflare/src/utils/handler.ts` maintenance and keep documentation simple and lean. Users should have one file to reach for customizing everything, including platform-specific entrypoints as long as they follow the request/response design as well as astro pipeline.

Reviewer 3: If the advanced routing entrypoint was configurable, it could point to the same file as wrangler's entrypoint. I wonder if that could work?

Reviewer 2: Interesting idea, however reading the RFC it sounds like the incoming `Request` inside the `fetch` handler of a provider-specific entrypoint is not the same as the one `app.ts` expects 🤔

Author: I stopped calling this "application entrypoint" because that's not accurate and can't be accurate. Astro can't get in front of adapters/platforms; those will always run first since they own the platform. If you create a worker.ts then that's runs first, regardless of what Astro does, that's `wrangler` who makes that decision.

But to address your concerns here, I think we can document that `astro/fetch` does work inside of `worker.ts` so you should be able to use that instead of `src/app.ts`. I think that does make sense. In which case we probably can deprecate `@astrojs/cloudflare/entrypoints/server` and tell people to just use `import { astro  } from 'astro/fetch';` instead since there's not much code between the two anyways.

Author: @florian-lefebvre No that doesn't work, the Cloudflare vite plugin loads `worker.ts`, Astro doesn't. The layer is always going to be:

(host runtime) -> (host entrypoint) -> (adapter) -> astro/app -> src/app.ts

We have no control over the first 3 things in the chain. That's why I moved away from calling this an "entrypoint". It's an entrypoint inside of _Astro's_ request handling (or as close as it can be).

Reviewer 2: If `astro/fetch` works inside of `worker.ts` that is great and we should definetly document it that way. In that case the decision for users would be the following in my opionon: 

If they use a custom entrypoint they should not use `src/app.ts`, if they don't use a custom entrypoint they can use `src/app.ts`

> we probably can deprecate @astrojs/cloudflare/entrypoints/server

That is my hope, but is that true though? The [handler](https://github.com/withastro/astro/blob/c0db63dd9659317e503dfce2d8eb50d6d67be34d/packages/integrations/cloudflare/src/utils/handler.ts#L37) from `@astrojs/cloudflare/entrypoints/server` does do some special stuff with `SESSION` & `ASSETS` binding and custom headers though, would that also work with `astro/fetch`? We would need to find a way to get that logic inside the functions of `astro/fetch`?

Reviewer 3: Is `app.ts` ever relevant for purely static sites? If not, that makes me wonder if this should be an adapter feature instead of core feature. Core could still provide `astro/(fetch|hono)` but it would only be used in adapter managed entrypoints

Author: Yes, redirects and i18n for example both work in static sites. Hono itself has a bunch as well, like sitemaps, and some development focused middleware.

Need to be careful to over-index on Cloudflare here, the Node adapter has no application entrypoint. Nor does the Vercel adapter.

Author: > That is my hope, but is that true though? The [handler](https://github.com/withastro/astro/blob/c0db63dd9659317e503dfce2d8eb50d6d67be34d/packages/integrations/cloudflare/src/utils/handler.ts#L37) from @astrojs/cloudflare/entrypoints/server does do some special stuff with SESSION & ASSETS binding and custom headers though, would that also work with astro/fetch? We would need to find a way to get that logic inside the functions of astro/fetch?

I'll look into this, but presumably we can replace this stuff with handlers, so if you're using Cloudflare you could do (as one example):

```ts
import { astro, FetchState } from 'astro/fetch';
import { cf } from '@astrojs/cloudflare/fetch';

export default {
  fetch(request: Request) {
    const state = new FetchState(request);
    await cf(state);
    return astro(state);
  }
}
```

Author: Here's a PR that adds these: https://github.com/withastro/astro/pull/16729

So you could skip creating a `src/app.ts` and put this in your worker instead, basically equivalent:

```ts
import { astro, FetchState } from 'astro/fetch';
import { cf } from '@astrojs/cloudflare/fetch';

export default {
  async fetch(request: Request, env: Env, ctx: ExecutionContext) {
    const state = new FetchState(request);
    const asset = await cf(state, env, ctx);
    if (asset) return asset;
    return astro(state);
  }
}
```

Reviewer 2: Looks nice. That would probably replace `import handler from "@astrojs/cloudflare/entrypoints/server";` in the future, which would be nice :)

`FetchState` is a mutable object that is created once per request and passed through all handlers. Handlers read and write properties on it as they run — for example, `redirects()` reads `routeData` to check for a match, and `pages()` sets `response` after rendering. It is not copied between handler calls; every handler receives the same instance.

The user creates the `FetchState` when using the `astro/fetch` API. The expected public surface includes:

- `request` - The current `Request` object.
- `response` - The `Response` produced by handlers, if any.
- `routeData` - The matched route information.
- `pathname` - The resolved pathname for the request. During a static build this may differ from the URL pathname — for example, it may include a `.html` extension to match the output file.
- `locals` - The `App.Locals` object, available to middleware and endpoints via `ctx.locals` / `Astro.locals`.
- `cookies` - The `AstroCookies` instance for reading/writing cookies.

> > **Note:** The exact shape of `FetchState` is still a work in progress and may change during the review and implementation process.

Reviewer 3: Did it change?

Author: Updated the list in 489e0be.

Usage:

```ts
import { FetchState, pages, redirects } from 'astro/fetch';

export default {
  fetch(request: Request) {
    const state = new FetchState(request);

    // Run the redirects
    const response = redirects(state);
    if(response) {
      return response;
    }

    // Continue...
    return pages(state);
  }
}
```

> Note that using the `astro/hono` API the creation of FetchState is not necessary, since Hono has its own Context object; handlers can get/create a FetchState by inspecting HonoContext for a certain key (likely `astro.fetchState`).

Reviewer 3: Is the key an implementation detail or should it be clearly documented?

Author: It's public, I'll open a PR to document, meanwhile here's the RFC change: 5520551

Author: Ah, you know what, it's not as easy as just documenting the key. it's created lazily on first access using the internal `getFetchState` function. We *could* expose this, but I'd rather wait until someone asks for it.

Reviewer 3: If I understand correctly, if there's no public API for it that means a package that wants to have handlers like the official ones can only implement some for use in a simple Fetchable, not in an Hono middleware. I think it should be a documented public API so both flavors have the same capabilities

Author: Updated the RFC:

- Implementation: https://github.com/withastro/astro/pull/16998
- Docs: https://github.com/withastro/docs/pull/14021

> ### Context Providers

Reviewer 3: I think this is fine for now but I don't really like the fact 3rd party handlers can extend the Astro global, there are `locals` for this exact usecase. So maybe the end game here is to remove this and move to locals, even for session and cache.

Again, not something to tackle now to avoid disruptions but to keep in mind for later.

Author: The idea of doing it this way was to make Astro more extensible, and eventually move things like Actions to a separate package, but still have the same API it has today. But as that is not happening now, I'd be fine with marking this API as internal and not supporting it for 3rd parties yet.

Reviewer 3: In that case I think it'd be better to keep it internal

Author: Done in https://github.com/withastro/astro/pull/16980

> ### Context Providers

Reviewer 3: I also want to note that this means eg. `Astro.session` may not be defined anymore if the user doesn't include it as a middleware, which can break integrations assuming it exists. So I think this could be considered breaking. What do you think?

Author: That wouldn't be breaking, the user didn't have the option of not including it before. It's still available by default, so `Astro.session` is defined.

Reviewer 3: I think it is breaking but maybe I misunderstand it:
- Before: `Astro.session` always defined
- After: If `fetchFile` is specified and the `session()` middleware is not used, then `Astro.session` is undefined. But the types say it is always defined

Is that correct?

(This is very much related to https://github.com/withastro/roadmap/pull/1344#discussion_r3362715698 probably)

Author: Breaking is when you don't change your code and something stops working. You have to change your code to get something that was previously working to work again. The only thing breaking in this RFC is that `src/fetch.ts` becomes a reserved file name. Everything else is opt-in.

Author: Actually maybe we are talking about different things. I'm thinking in terms of semver `major` breaking changes. Are you just pointing out that this breaks assumptions of integrations? 

If so, I could add a note to a section on Caveats, etc.

Author: I'm pretty sure that's what you're saying here. In that case I think what I'd rather do is just provide stubs so that Astro.session continues to exist even if the handler is not added.

Reviewer 3: Yes I'm talking from the POV of the integrations. I think section in caveats + mention in docs + stub would be great if you don't mind!

Author: So the situation is not what we thought, `Astro.session` was always possibly undefined. It changed from an optional prop in 5.x: https://github.com/withastro/astro/blob/astro%405.18.0/packages/astro/src/types/public/context.ts

to `AstroSession | undefined` now: https://github.com/withastro/astro/blob/main/packages/astro/src/types/public/context.ts

So the only difference now is that there isn't a getter with a clear error message. I'll add that back. But there isn't a breaking change from the integration author perspective, they always needed to check for it existing.

Author: https://github.com/withastro/astro/pull/16982

- `finalizeAll(): Promise<void> | void` — Runs all registered `finalize` callbacks for providers that were actually resolved. Returns `void` synchronously when nothing needs finalizing; returns a `Promise<void>` when any `finalize` callback is asynchronous. Callers should `await` the result.

For example, the `sessions()` handler registers a provider that lazily creates an `AstroSession` and persists it on finalize:

```ts
state.provide<AstroSession>('session', {
  create() {
    return new AstroSession({ cookies, config, ... });
  },
  finalize(session) {
    return session.persist();
  },
});
```

From the user's perspective they only need to add the `session()` handler/middleware and then they get `ctx.session` / `Astro.session` as values. This could also be used by 3rd party integrations.

### Typing Context Providers

Custom context providers are typed using the same pattern as `locals`: module augmentation. A library or user registers a provider with a string key, and the corresponding type is declared by extending an interface in `env.d.ts`:

```ts
declare namespace App {
  interface Providers {
    oauth: import('./lib/oauth').OAuthSession;
  }
}
```

> With this declaration, `ctx.oauth` and `Astro.oauth` are properly typed. Built-in providers like `session` are pre-declared by Astro's own type definitions, so users don't need to augment those themselves.

Reviewer 3: That means `Astro.session` will be typed even if the corresponding middleware isn't set up. Is there anything that can be done to improve this?

Author: I don't think so... unfortunately types need to be static. I don't think we could even use injectTypes here since we don't know its not used until much later. I think ensuring we have good runtime errors is probably the best we can do.

## Separate `@astrojs/hono` package

An alternative to `astro/hono` would be a separate `@astrojs/hono` package. However, Astro does not depend on Hono at runtime — the user brings their own Hono dependency. The `astro/hono` module only provides the thin adapter layer that bridges Astro's feature handlers into Hono middleware. Since there is no Hono code being bundled into Astro, a separate package is unnecessary; a subpath export is sufficient.

# Adoption strategy

The `src/app.ts` module will be opt-in. Without this file Astro will behave as it normally does.

Initially this feature will be gated by an `experimental.advancedRouting` flag. In the next major version of Astro there will be no flag, meaning that `src/app.ts` becomes a special-file that users can't use for other purposes.

To accommodate projects which already have a `app.ts` file, this will be configurable:

```
export default defineConfig({

>   fetchFile: 'fetch.ts'

Reviewer 3: I think it should be called `fetchEntrypoint` to match other APIs, and allow:
- relative paths as strings
- package specifiers
- URLs

Also I don't know if `fetchX` is the right name but I don't have a better idea and I don't want to derail this comment ;)

Author: I don't think it should allow package specifiers. The primary goal of this RFC is to allow user to have control over routing. So the intent is for this to be a file in the project. It's an explicit non-goal of allowing integrations to "take over" this feature and I think allowing package specifiers would encourage that. The way for 3rd parties to participate is to export handlers/middleware that the user can user.

For this reason I want to keep it as `fetchFile`. It's similar to middleware, actions, and content config, all of which are name files (and not configurable at all).
